### ISSUE TITLE
Various versions

### ISSUE THREAD

**Issue by nicehashdev** [tromp/equihash#3]
I see you have various versions; dev, eq etc... some clarification on these would be great.

But overall, great work and thanks. Working on implementing CUDA version into a miner now.

**Comment by tromp** [comment]
> I see you have various versions; dev, eq etc... some clarification on these would be great.

Those are various development versions, that may or may not work at any time.
Once I reach a non-trivial speed-up with them I'll move them to the
equi_miner...

So just ignore them, unless you're curious to see what I'm experimenting with...

-John

**Comment by nicehashdev** [comment]
Ah ok, thanks for explanation. So.. the best would be to use equi_miner.cu for CUDA? That one I used and it seems like it is working fine (validation was OK) for most of thread settings (4096 threads gave out different result - 0 solutions).

**Comment by tromp** [comment]
Yes, use equi_mine.cu for CUDA. For newly compiles executables, it always seems to get 0 solutions on the first run; I don't know why.

**Comment by jonesskill01** [comment]
Tromp, how to run this miner on ubuntu comand line?
Im trying to run equi_mine.cu but the output are: command not find

**Comment by tromp** [comment]
make eqcuda will make the cuda executable.
assuming you have the CUDA developer tools installed,
including the nvcc compiler...

regards,
-John

**Comment by lfarkas** [comment]
what is the current state of these algorithm? and how can i integrate it into official zcashd? it'd be useful to be a few lines of howto for this integration. is there any plane to include it into upstream zcashd?

**Comment by tromp** [comment]
What do you mean by state?
An earlier version of equi1 is already integrated in zcashd,
thanks to work by sarath-hotspot, str4d, and daira.
See https://github.com/zcash/zcash/tree/master/src/pow/tromp

**Comment by lfarkas** [comment]
i mean do you further refine them?

will both the cuda and the cpu version will be integrated? 
is there any PR for this? or is there any reason why not integrated?

**Comment by tromp** [comment]
I currently have no ideas for further improvement.
The cuda version will likely not be integrated into zcashd,
due to dependencies and stability issues.

**Comment by lfarkas** [comment]
it can be added as a plugin or conditionally compiled algorithm.

**Comment by tromp** [comment]
The CUDA solver could indeed be added to str4d's stand-alone miner.
But right now I have some other things to work on...

**Comment by lfarkas** [comment]
a small patch to the current version

```
diff --git a/equi.h b/equi.h
index 6dcb8d7..de57cba 100644
--- a/equi.h
+++ b/equi.h
@@ -67,7 +67,7 @@ void setheader(blake2b_state *ctx, const char *headernonce) {
 enum verify_code { POW_OK, POW_HEADER_LENGTH, POW_DUPLICATE, POW_OUT_OF_ORDER, POW_NONZERO_XOR };
 const char *errstr[] = { "OK", "wrong header length", "duplicate index", "indices out of order", "nonzero xor" };

-void genhash(blake2b_state *ctx, u32 idx, uchar *hash) {
+void genhash(const blake2b_state *ctx, u32 idx, uchar *hash) {
   blake2b_state state = *ctx;
   u32 leb = htole32(idx / HASHESPERBLAKE);
   blake2b_update(&state, (uchar *)&leb, sizeof(u32));
@@ -76,7 +76,7 @@ void genhash(blake2b_state *ctx, u32 idx, uchar *hash) {
   memcpy(hash, blakehash + (idx % HASHESPERBLAKE) * WN/8, WN/8);
 }

-int verifyrec(blake2b_state *ctx, u32 *indices, uchar *hash, int r) {
+int verifyrec(const blake2b_state *ctx, u32 *indices, uchar *hash, int r) {
   if (r == 0) {
     genhash(ctx, *indices, hash);
     return POW_OK;
diff --git a/equi_miner.h b/equi_miner.h
index da4b7dd..252afdf 100644
--- a/equi_miner.h
+++ b/equi_miner.h
@@ -36,6 +36,7 @@

 #include "equi.h"
 #include <stdio.h>
+#include <stdlib.h>
 #include <pthread.h>
 #include <assert.h>

```

**Comment by tromp** [comment]
Thx for the const reminders.
stdlib.h is already included from equi.h ...

**Comment by lfarkas** [comment]
one more note in equi.h you add a new function setheader. unfortunately this use blake2b_param and blake2b_init_param which are internal struct and function of sodium so it's not possible to use if one would like to use your hash with sodium (like zcash like to use it).

**Comment by tromp** [comment]
dear Levente,

> one more note in equi.h you add a new function setheader. unfortunately this use blake2b_param and blake2b_init_param which are internal struct and function of sodium so it's not possible to use if one would like to use your hash with sodium (like zcash like to use it).

I used the reference implementation of blake
https://github.com/BLAKE2/BLAKE2
which doesn't have the
crypto_generichash_blake2b_init_salt_personal()
that https://github.com/zcash/zcash/blob/master/src/crypto/equihash.cpp
uses. Unlike zcash, I didn't use anything sodium-specific.

regards,
-John

**Comment by lfarkas** [comment]
FYI:
https://github.com/zcash/zcash/pull/1788

**Comment by tromp** [comment]
> FYI:
> zcash/zcash#1788

Thanks for preparing this, Levente!
Note that this leaves the AVX2 optimizations unused.
Support for that would require adding the blake2-avx subdirectory,
building an extra target, and allowing both "tromp" and "tromp2"
as values for equihashsolver in zcash.conf.

regards,
-John

**Comment by lfarkas** [comment]
huu.
blake2-avx2 or blake2-asm or both?
actually i don't understand the algorithm itself just port it to zcash. anyway it's not the same as libsodium's avx2 optimization?
https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_generichash/blake2/ref/blake2b-compress-avx2.h
they automatically detect processor type.

anyway if i port avx2 to zcash too which version to choose? AFAIS "8 x eqavx21: 20.3 Sol/s" seems to be the fastest. 
so only blake2-avx2 directory needed? it seems you never use any asm file in this case. is it true?
may be i open a new issue in zcash.

**Comment by tromp** [comment]
dear Levente,

> blake2-avx2 or blake2-asm or both?

My make target eqavx21 needs only blake2-avx2.
blake2-asm is for dev1 which I don't expect to be integrated into zcashd
as it relies on a particular assembler.

> actually i don't understand the algorithm itself just port it to zcash. anyway it's not the same as libsodium's avx2 optimization?

It's a non-standard optimization, as it computes 4 independent
blake2b's in parallel.

> https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_generichash/blake2/ref/blake2b-compress-avx2.h
> they automatically detect processor type.

I haven't had time to study that in detail, but in any case I think a separate
tromp2 executable is fine. It also simplifies the discussion of its performance.

> anyway if i port avx2 to zcash too which version to choose? AFAIS "8 x eqavx21: 20.3 Sol/s" seems to be the fastest.

zcashd only runs single-threaded versions, so eqavx21 is the right
one. Users will set genproclimit to choose how many instances to run.

> so only blake2-avx2 directory needed? it seems you never use any asm file in this case. is it true?

Correct. Thanks again for all your effforts.

regards,
-John

**Comment by lfarkas** [comment]
On 11/05/2016 04:55 PM, John Tromp wrote:

> > https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_generichash/blake2/ref/blake2b-compress-avx2.h
> > they automatically detect processor type.
> 
> I haven't had time to study that in detail, but in any case I think a
> separate
> tromp2 executable is fine. It also simplifies the discussion of its
> performance.

what do you mean by "tromp2 executable" in case of zcashd there is only
one executable which should have to contain both tromp and tromp2
equihashsolver.

## 

  Levente                               "Si vis pacem para bellum!"

**Comment by tromp** [comment]
> what do you mean by "tromp2 executable" in case of zcashd there is only
> one executable which should have to contain both tromp and tromp2
> equihashsolver.

You're right; I was confused about how the solver choice is implemented.
So we need either runtime detection of avx2 support, in the equi() constructor,
or compile time detection with #ifdef **AVX2** to decide which blake2b
implementation to use.
